File: csharp-dotnet/src/Architecture.Core/Functional/Result.cs
using System;

namespace Architecture.Core.Functional;

/// <summary>
/// Represents the result of an operation that can either succeed or fail without returning a value.
/// Provides monadic operations for functional error handling and composition.
/// </summary>
public readonly struct Result : IEquatable<Result>
{
    private readonly bool _isSuccess;
    private readonly Error _error;

    /// <summary>
    /// Gets a value indicating whether the operation succeeded.
    /// </summary>
    public bool IsSuccess => _isSuccess;

    /// <summary>
    /// Gets a value indicating whether the operation failed.
    /// </summary>
    public bool IsFailure => !_isSuccess;

    /// <summary>
    /// Gets the error information if the operation failed.
    /// Throws InvalidOperationException if accessed when IsSuccess is true.
    /// </summary>
    public Error Error
    {
        get
        {
            if (_isSuccess)
                throw new InvalidOperationException("Cannot access Error when Result is successful");
            return _error;
        }
    }

    private Result(bool isSuccess, Error error)
    {
        _isSuccess = isSuccess;
        _error = error;
    }

    /// <summary>
    /// Creates a successful result.
    /// </summary>
    /// <returns>A successful Result instance.</returns>
    public static Result Ok() => new(true, default);

    /// <summary>
    /// Creates a failed result with the specified error.
    /// </summary>
    /// <param name="error">The error that caused the failure.</param>
    /// <returns>A failed Result instance.</returns>
    public static Result Fail(Error error) => new(false, error);

    /// <summary>
    /// Transforms a successful result by applying a function that returns a value.
    /// If the result is failed, returns a failed Result&lt;T&gt; with the same error.
    /// </summary>
    /// <typeparam name="T">The type of value to return.</typeparam>
    /// <param name="func">The function to apply if successful.</param>
    /// <returns>A Result&lt;T&gt; containing the transformed value or the original error.</returns>
    public Result<T> Map<T>(Func<T> func)
    {
        ArgumentNullException.ThrowIfNull(func);

        return _isSuccess
            ? Result<T>.Ok(func())
            : Result<T>.Fail(_error);
    }

    /// <summary>
    /// Monadic bind operation that chains operations that can fail.
    /// If this result is successful, executes the function; otherwise returns the error.
    /// </summary>
    /// <param name="func">The function to execute if this result is successful.</param>
    /// <returns>The result of the function or the original error.</returns>
    public Result Bind(Func<Result> func)
    {
        ArgumentNullException.ThrowIfNull(func);

        return _isSuccess ? func() : this;
    }

    /// <summary>
    /// Pattern matching method that executes one of two functions based on success/failure.
    /// </summary>
    /// <typeparam name="T">The type of value to return.</typeparam>
    /// <param name="onSuccess">Function to execute if successful.</param>
    /// <param name="onFailure">Function to execute if failed.</param>
    /// <returns>The result of the executed function.</returns>
    public T Match<T>(Func<T> onSuccess, Func<Error, T> onFailure)
    {
        ArgumentNullException.ThrowIfNull(onSuccess);
        ArgumentNullException.ThrowIfNull(onFailure);

        return _isSuccess ? onSuccess() : onFailure(_error);
    }

    /// <summary>
    /// Implicitly converts an Error to a failed Result.
    /// </summary>
    /// <param name="error">The error to convert.</param>
    public static implicit operator Result(Error error) => Fail(error);

    /// <summary>
    /// Converts an Error to a failed Result.
    /// </summary>
    /// <param name="error">The error to convert.</param>
    /// <returns>A failed Result.</returns>
    public static Result FromError(Error error) => Fail(error);

    /// <summary>
    /// Determines whether two Result instances are equal.
    /// </summary>
    public bool Equals(Result other)
    {
        if (_isSuccess != other._isSuccess) return false;
        return _isSuccess || _error.Equals(other._error);
    }

    /// <summary>
    /// Determines whether this Result is equal to another object.
    /// </summary>
    public override bool Equals(object? obj) => obj is Result other && Equals(other);

    /// <summary>
    /// Gets the hash code for this Result.
    /// </summary>
    public override int GetHashCode() => _isSuccess ? 1 : _error.GetHashCode();

    /// <summary>
    /// Returns a string representation of this Result.
    /// </summary>
    public override string ToString() => _isSuccess ? "Success" : $"Failure: {_error}";

    /// <summary>
    /// Determines whether two Results are equal.
    /// </summary>
    public static bool operator ==(Result left, Result right) => left.Equals(right);

    /// <summary>
    /// Determines whether two Results are not equal.
    /// </summary>
    public static bool operator !=(Result left, Result right) => !left.Equals(right);
}

File: java-spring/architecture-core/src/main/java/com/architecture/core/functional/Result.java
package com.architecture.core.functional;

import java.util.Objects;
import java.util.function.Function;

/**
 * Represents the result of an operation that can either succeed with a value or fail with an error.
 * Implements monadic operations for functional composition and error handling.
 *
 * Requirements: FR-008, FR-011, FR-013 - Result with monadic operations
 *
 * @param <T> The type of the success value
 */
public final class Result<T> {
    private final T value;
    private final Error error;
    private final boolean isSuccess;

    private Result(T value, Error error, boolean isSuccess) {
        this.value = value;
        this.error = error;
        this.isSuccess = isSuccess;
    }

    /**
     * Creates a successful result with the given value.
     *
     * @param value The success value (cannot be null)
     * @param <T>   The type of the value
     * @return A successful Result
     */
    public static <T> Result<T> success(T value) {
        Objects.requireNonNull(value, "Success value cannot be null");
        return new Result<>(value, null, true);
    }

    /**
     * Creates a successful result with no value (for Void operations).
     *
     * @return A successful Result<Void>
     */
    public static Result<Void> success() {
        return new Result<>(null, null, true);
    }

    /**
     * Creates a failed result with the given error.
     *
     * @param error The error (cannot be null)
     * @param <T>   The type of the would-be success value
     * @return A failed Result
     */
    public static <T> Result<T> failure(Error error) {
        Objects.requireNonNull(error, "Error cannot be null");
        return new Result<>(null, error, false);
    }

    /**
     * Creates a failed result with the given code and message as a domain error.
     *
     * @param code    The error code
     * @param message The error message
     * @param <T>     The type of the would-be success value
     * @return A failed Result with a domain error
     */
    public static <T> Result<T> failure(String code, String message) {
        return failure(Error.domain(code, message));
    }

    /**
     * Checks if this result represents a success.
     *
     * @return true if this is a successful result
     */
    public boolean isSuccess() {
        return isSuccess;
    }

    /**
     * Checks if this result represents a failure.
     *
     * @return true if this is a failed result
     */
    public boolean isFailure() {
        return !isSuccess;
    }

    /**
     * Gets the success value.
     *
     * @return The success value
     * @throws IllegalStateException if this is a failed result
     */
    public T getValue() {
        if (!isSuccess) {
            throw new IllegalStateException("Cannot get value from failed result");
        }
        return value;
    }

    /**
     * Gets the error.
     *
     * @return The error
     * @throws IllegalStateException if this is a successful result
     */
    public Error getError() {
        if (isSuccess) {
            throw new IllegalStateException("Cannot get error from successful result");
        }
        return error;
    }

    /**
     * Maps the success value to a new type using the provided mapper function.
     * If this is a failure, returns a failure with the same error.
     *
     * @param mapper The mapping function
     * @param <U>    The type of the mapped value
     * @return A Result containing the mapped value or the original error
     */
    public <U> Result<U> map(Function<? super T, ? extends U> mapper) {
        Objects.requireNonNull(mapper, "Mapper function cannot be null");
        if (isSuccess()) {
            try {
                return Result.success(mapper.apply(getValue()));
            } catch (Exception e) {
                return Result.failure(Error.infrastructure("Mapping.Failed", e.getMessage(), e));
            }
        }
        return Result.failure(getError());
    }

    /**
     * Binds this result to another result-producing function.
     * If this is a failure, returns a failure with the same error.
     *
     * @param binder The binding function
     * @param <U>    The type of the bound result value
     * @return The result of the binding operation or the original error
     */
    public <U> Result<U> bind(Function<? super T, Result<U>> binder) {
        Objects.requireNonNull(binder, "Binder function cannot be null");
        if (isSuccess()) {
            try {
                return binder.apply(getValue());
            } catch (Exception e) {
                return Result.failure(Error.infrastructure("Binding.Failed", e.getMessage(), e));
            }
        }
        return Result.failure(getError());
    }

    /**
     * Matches this result against success and failure handlers.
     *
     * @param onSuccess The function to apply if this is a success
     * @param onFailure The function to apply if this is a failure
     * @param <U>       The return type of both handlers
     * @return The result of applying the appropriate handler
     */
    public <U> U match(Function<? super T, ? extends U> onSuccess, Function<Error, ? extends U> onFailure) {
        Objects.requireNonNull(onSuccess, "Success handler cannot be null");
        Objects.requireNonNull(onFailure, "Failure handler cannot be null");
        return isSuccess() ? onSuccess.apply(getValue()) : onFailure.apply(getError());
    }

    /**
     * Gets the success value or returns the default value if this is a failure.
     *
     * @param defaultValue The default value to return if this is a failure
     * @return The success value or the default value
     */
    public T getValueOrDefault(T defaultValue) {
        return isSuccess() ? getValue() : defaultValue;
    }

    /**
     * Gets the success value or throws a ResultException if this is a failure.
     *
     * @return The success value
     * @throws ResultException if this is a failure
     */
    public T getValueOrThrow() {
        if (isSuccess()) {
            return getValue();
        }
        throw new ResultException(getError());
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Result<?> result = (Result<?>) obj;
        return isSuccess == result.isSuccess &&
               Objects.equals(value, result.value) &&
               Objects.equals(error, result.error);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, error, isSuccess);
    }

    @Override
    public String toString() {
        return isSuccess()
            ? String.format("Success{value=%s}", value)
            : String.format("Failure{error=%s}", error);
    }
}